### DISCUSSION TITLE
Tree-View Layout

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#3054] [Feature Requests]
If you have deeply nested (or self-nested) items, it would be nice to have a layout that let's you more efficiently browse, see, explore those items by expanding each item and seeing the children. In this way you could have a nice tree-view that async loads relational content.

I can build a design for this if someone's interested in taking it on!

**Comment by danieljwestman** [comment]
This would indeed be sweet!

**Comment by Sennik** [comment]
I used this approach for years in my CMS (Which I love to replace with Directus). 

It is IMHO necessary to have an option to create hierarchical content for a website to begin with. It is very usefull to create and maintain navigation without having to create them seperately and connect content to the nodes.

**Reply by rijkvanzanten** [reply]
> It is IMHO necessary to have an option to create hierarchical content for a website to begin with

Directus does have full relational fields support already of course, this would just be an extra way to view nested data in a more hierarchical structure on the listing pages 🙂

**Reply by Sennik** [reply]
Indeed, but not only to view, but also to create and change I assume. I really think it would be a very helpful addition to the existing layouts. Usen tree view for webcontent (in menu structures), and table view for generic content like news, or projects or blogs, etc

**Reply by everyx** [reply]
@rijkvanzanten  In tree view, if we want add the ability to change a node's depth, parent, and order(maybe drag and drop), we need to have a set rule of the data struct of the tree, something like this [Nested set model](https://en.wikipedia.org/wiki/Nested_set_model), looking forward to this feature :)

**Reply by cupcakearmy** [reply]
Can second that, almost any website needs this. It's possible without it as @rijkvanzanten said but it would improve the product by a ton :)

**Reply by Sennik** [reply]
To be honest, I am waiting for this to start even using Directus as it is useless form without.

**Comment by bcartier** [comment]
Would love to see this feature in the core product!

**Reply by benhaynes** [reply]
It's a difficult one, unless all the data (parent and children) will come from the same collection. That said, we're also looking into a way to show nested relational data on ALL Layouts (eg; a folder structure prepended above the Layout.

I believe a Spreadsheet Layout might beat this one to production though.

**Reply by Sennik** [reply]
Data from one collection will do for me ;-)

**Reply by eXaminator** [reply]
It feels like allowing a single collection to be nested and have an interactive tree view (just like we have table, cards etc. now) would probably help solve many use cases (the most common one would probably nested pages / website navigation).

**Reply by rijkvanzanten** [reply]
@eXaminator Website navigation is an interesting one though. Often times, you'd want to differentiate between `menus` and `menu_items`, in which case it's no longer a single recursive type 🤔

**Reply by LasseRosenow** [reply]
@rijkvanzanten Having `TreeViews` for more then just single type collections would be great 😀 I wrote down a Feature request as to how it could be achieved a while ago: #7447

(The technical details are within our conversation below the main post)

**Reply by eXaminator** [reply]
So my use case is specifically intended to be simple. I don't want menus and menu items. I just want pages. Root pages will be displayed in the main navigation, children in a second or third level navigation. If I wanted a deviation between menus and menu items, I could probably have a workaround using rules to change the relevant fields based on wether it's a root element or not. I think this could be a useful first step. Of course, allowing any kind of tree would be neat, but that could easily become a UX hell I guess...

**Reply by rijkvanzanten** [reply]
>  but that could easily become a UX hell I guess...

Haha yeah exactly! That's what I'm trying to think through here 🙂

**Comment by MMTHD** [comment]
"Upvote" for rethink the priority of implementation of this feature (from Q3 2022 to Q3 2021 :D)

**Reply by MMTHD** [reply]
My thoughts about the functionality of this is following:

- By creating a collection we should have a checkbox "Nested/Hierarchical", which creates automatically some fields which are necessary for that like child, parent, sort etc.
- In the frontend view, it should be the same table like now but the draggable option should be able to nesting items (https://sortablejs.github.io/Vue.Draggable/#/nested-with-vmodel)

I think the implementation should be simple, do i forgot some things?

**Reply by Sennik** [reply]
Exactly. Or simply drag an item a couple of mm to the right to 'create' the parent/child relationship, like Statamic offers. Statamic uses no database though (until you need one ;-))

**Comment by KrisKnez** [comment]
I think this is a great idea. We should have this!

**Comment by VaZark** [comment]
I’m not super well versed in db hierarchy logic but maybe we can take some inspiration from [django-treebeard](https://github.com/django-treebeard/django-treebeard)  ?

**Comment by KrisKnez** [comment]
I'd like to note that the Tree-view layout is in the roadmap.

![image](https://user-images.githubusercontent.com/11775989/155032599-2376cb6e-9208-4193-9d11-c29fa7af79fd.png)
Hopefully, it is on schedule! 😄

**Reply by VaZark** [reply]
Sql views too!! That makes life so much easier when needing denormed data

**Comment by brett-james-rocketlab** [comment]
Is there any kind of inprogress screenshot if it's available, to give us an idea of how it's planned to look like? Also keen to make use of this request.

**Comment by cloudcompute** [comment]
Hello, Using this tree, would it be possible to see our website content in the form of a tree.. and the user would be able to edit the content of a particular node by clicking on it?

**Comment by valery** [comment]
Hi, Any progress to show for this so much needed feature?

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by rijkvanzanten** [comment]
New RFC based take on the same thing: https://github.com/directus/directus/discussions/18651 🙂  (cc @w0kyj)

**Comment by w0kyj** [comment]
Adding RFC spec format and aggregated notes from other tickets:
Other related GH Discussions:
https://github.com/directus/directus/discussions/7447
https://github.com/directus/directus/discussions/18651

# **Summary**

A tree view layout adds presentation as hierarchical list of deeply nested (or self-nested) items. The layout would allow the content user to more efficiently browse and manage parent child relational data. 

# **Basic Examples**

Tree views can be used to represent:

- hierarchical structure representing some engineering locations, called “functional locations”
    - each location may have sub-locations via recursive tree relationship

![tree-example-1](https://github.com/directus/directus/assets/40695629/6ca2b5aa-4358-4912-985a-d09d24cadda0)
    
![tree-example-2](https://github.com/directus/directus/assets/40695629/a1455118-03d6-498d-80b5-f8932d582c46)

    
- parent/child pages
    - an item representing a web page with a recursive tree relationship may be expanded to show the related sub-pages
- project management displaying the tasks and recursive sub-tasks

# **Motivation**

Improve user experience and understanding of Tree hierarchy fields within collections. Currently user can only view the trees hierarchy in the individual item form.

This is beneficial for a wide variety of use cases with hierarchical relationships including files/folders web pages, menus, project management, inventory, locations, etc

# **Detailed Design**

Support relational type: tree view

- Allow manual sorting of tree items in the layout
    - needed on all displayed levels:

## Column Sorting Examples

1. **Initial table**

| Level0 | Level1 | Level2 | Text | Number |
| --- | --- | --- | --- | --- |
| A |  |  | Parent A | 45 |
|  | AA |  | Child AA | 42 |
|  |  | AAA | Sub-child AAA | 154 |
|  |  | AAB | Sub-child AAB | 11 |
|  | AB |  | Child AB | 37 |
|  |  | ABA | Sub-child ABA | 12 |
| B |  |  | Parent B | 99 |
|  | BA |  | Child BA | 0 |
1. **Hierarchy Column sorted descending by Level0**

| Level0 | Level1 | Level2 | Text | Number |
| --- | --- | --- | --- | --- |
| B |  |  | Parent B | 99 |
|  | BA |  | Child BA | 0 |
| A |  |  | Parent A | 45 |
|  | AA |  | Child AA | 42 |
|  |  | AAA | Sub-child AAA | 154 |
|  |  | AAB | Sub-child AAB | 11 |
|  | AB |  | Child AB | 37 |
|  |  | ABA | Sub-child ABA | 12 |
1. **Hierarchy Column sorted descending by Level1**

| Level0 | Level1 | Level2 | Text | Number |
| --- | --- | --- | --- | --- |
| A |  |  | Parent A | 45 |
|  | AB |  | Child AB | 37 |
|  |  | ABA | Sub-child ABA | 12 |
|  | AA |  | Child AA | 42 |
|  |  | AAA | Sub-child AAA | 154 |
|  |  | AAB | Sub-child AAB | 11 |
| B |  |  | Parent B | 99 |
|  | BA |  | Child BA | 0 |
1. **Non-Hierarchy Column sorted ascending**

| Level0 | Level1 | Level2 | Text | Number |
| --- | --- | --- | --- | --- |
| A |  |  | Parent A | 45 |
|  | AB |  | Child AB | 37 |
|  |  | ABA | Sub-child ABA | 12 |
|  | AA |  | Child AA | 42 |
|  |  | AAB | Sub-child AAB | 11 |
|  |  | AAA | Sub-child AAA | 154 |
| B |  |  | Parent B | 99 |
|  | BA |  | Child BA | 0 |

## Item **Manual Sorting**

1. **Hierarchy Change by Level0**

| Manual Sort | Level0 | Level1 | Level2 | Text | Number |
| --- | --- | --- | --- | --- | --- |
| == | B |  |  | Parent B | 99 |
| == |  | BA |  | Child BA | 0 |
| == | A |  |  | Parent A | 45 |
| == |  | AA |  | Child AA | 42 |
| == |  |  | AAA | Sub-child AAA | 154 |
| == |  |  | AAB | Sub-child AAB | 11 |
| == |  | AB |  | Child AB | 37 |
| == |  |  | ABA | Sub-child ABA | 12 |
1. **OPTIONAL**: Child Hierarchy Changes (Level1+)

| Manual Sort | Level0 | Level1 | Level2 | Text | Number |
| --- | --- | --- | --- | --- | --- |
| == | A |  |  | Parent A | 45 |
| == |  | AB |  | Child AB | 37 |
| == |  |  | ABA | Sub-child ABA | 12 |
| == |  | AA |  | Child AA | 42 |
| == |  |  | AAA | Sub-child AAA | 154 |
| == |  |  | AAB | Sub-child AAB | 11 |
| == | B |  |  | Parent B | 99 |
| == |  | BA |  | Child BA | 0 |

| Manual Sort | Level0 | Level1 | Level2 | Text | Number |
| --- | --- | --- | --- | --- | --- |
| == | A |  |  | Parent A | 45 |
| == |  | AA |  | Child AA | 42 |
| == |  |  | AAA | Sub-child AAA | 154 |
| == |  |  | AAB | Sub-child AAB | 11 |
| == | AB |  |  | Parent AB | 37 |
| == |  | ABA |  | Child ABA | 12 |
| == | B |  |  | Parent B | 99 |
| == |  | BA |  | Child BA | 0 |

## Changing Root Item Option

<img width="768" alt="tree-example-file-3" src="https://github.com/directus/directus/assets/40695629/3dc187a7-4a08-49e0-acc8-bb2905a00581">


## Requirements List

### Must haves:

- [ ]  Same features and capabilities of the current “Table” layout with added support for Tree View fields to show nested hierarchy of the children
    - [ ]  Existing table layout + to add columns
        - [ ]  NOTE: Parent and child have same columns for Tree View fields
    - [ ]  In case ideas/rules needed for the tree views, resources:
        - [ ]  [Nested set model](https://en.wikipedia.org/wiki/Nested_set_model)
        - [ ]  https://github.com/django-treebeard/django-treebeard
        - [ ]  In the frontend view, it should be the same table like now but the draggable option should be able to nesting items (https://sortablejs.github.io/Vue.Draggable/#/nested-with-vmodel)
    - [ ]  Root Manual Sorting - Example above
- [ ]  Column Sorting
    - [ ]  All fields, but the hierarchy fields would be prioritized
        - [ ]  Examples above
- [ ]  Layout Configuration
    - [ ]  Hierarchy Source Field
        - [ ]  Only one hierarchy field at a time
        - [ ]  Ideally this is a filtered list of valid relational fields (similar to Kanban configurations)

### Should haves:

- [ ]  If the tree is expanded and you click on any of the child items, then open the clicked item

### Could haves:

- [ ]  Tree View Fields
    - [ ]  Manual Sorting for Child Hierarchy (Level1+)
        - [ ]  Change parent, depth and/or order
        - [ ]  Examples above
- [ ]  Syncs opened hierarchies to local storage
    - [ ]  App tries to return to same view when returning from item edits, etc.
- [ ]  Some way to select/click/double-click child item and make it the new current root for deeper viewing
    - [ ]  Similar to File Folder explorers on Operating Systems
    - [ ]  Add example of MacOS file explorer - JONATHAN

### Will not have:

- [ ]  Support for other relational types: m2o to another table, m2m(translations), m2a, file, files, images

# **Drawbacks**

Performance overhead for the table layout

# **Alternatives**

Custom layout extension built by clients or partners.

# **Adoption Strategy**

Augments existing layouts. Users may choose to use or not. All clients with recursive tree relationships will be able to visualize the parent-child relationships more easily.

# **Unresolved Questions**

**Additional Relationship Support Considerations**

- Display of parent and child field(s) for other relationships (m2o, o2m, m2m(translations), m2a, file, files, images)
    - Only a single column -- use Collection Display for each level?
    - Single column -- layout configuration Display for each level (max number of levels)?
    - Multiple columns - if this, then need detailed requirements, considerations include
        - What if there are multiple relational fields?
        - What if the child has additional relational fields?
- How many levels should be allowed in the display?
    - Configurable in the layout configuration?
    - Env or Hard max limit?
    - If multiple columns, depth enforcement is different for children vs parents?
- App Queries & Caching
    - Starts closed for all root items - performance issues for any other alternatives
    - How much data should be fetched initially?
    - When to fetch child level data?
- Pagination
    - What happens when the child rows exceed the number of page rows?
    -

**Reply by eXaminator** [reply]
I would like to propose a "could have", even though it might not strictly be part of the table view itself but also of other tree components: the ability to define a maximum depth (number of levels). Maybe this could be a feature of just the tree view, in that it would only show items up to a certain level and would only allow interaction with them.

Another "could have": allow the creation of a new child elements directly from a tree view, which would open the normal creation form but with a prepopulated parent field.

**Reply by rijkvanzanten** [reply]
I think that first could have is actually more of a must-have—or at least we need to have an upper limit to make sure we can keep the app performant

**Comment by formfcw** [comment]
I have created a design prototype for a tree structure where the main idea is to integrate the feature into the existing table layout rather than adding a new one. Have a look:

https://github.com/directus/directus/assets/78852214/4a2d6c4d-b020-4a66-b496-3b999b3449d3

**Reply by madc** [reply]
This looks amazing!!

**Reply by LasseRosenow** [reply]
I love this!!!

**Reply by danieljwestman** [reply]
Perfection

**Reply by formfcw** [reply]
Thank you guys :)

**Reply by benhaynes** [reply]
*Really* love this... awesome mockup!! CC @rijkvanzanten

**Reply by formfcw** [reply]
Thank you, nice to hear!

**Reply by vitalijalbu** [reply]
hope we'll get it one day... 😁

**Reply by danieljwestman** [reply]
Is this something that is actively being worked on? Are you building this as extension @formfcw ? 🙌🏼

**Reply by formfcw** [reply]
> Is this something that is actively being worked on? Are you building this as extension @formfcw ? 🙌🏼

No, I am not working on an extension.

**Reply by phil-sola** [reply]
Is this feature currently under development? I can see from the above screenshot of the roadmap it was on their in 2022 and looks like it's under 'next' on the current roadmap. 

Tree view for hierarchical content under collections, is essential really for understanding at a quick glance the relationships between pages and etc.

I would love to see this get released!

**Reply by guidomallee** [reply]
This is a feature that much more straightforward CMSs (such as DatoCMS) already have for a long time.

Is there any update on this feature making it to Directus?

The Tree View field doesn't do it as it lacks sorting, and this is not something you want to be managing on record level, but rather from a global view.

**Comment by tonnyavery** [comment]
Not sure how related is it to this discussion, hope it is.
I have built a collection with Tree View setup, that is essentially a tree of nested categories. Tree is quite deep.
Now when I connect my `categories` as many-to-many relationship to a `product` table, I was expecting to see a tree-view there, but instead I see just a table that is very difficult to manage and work with.
Would this new "Tree View Layout" allow to better manage such nested collections? If yes, any update on the progress?

<img width="375" alt="image" src="https://github.com/directus/directus/assets/1481332/0ed5049b-192e-43e0-9f9f-01ed5f23efeb">

**Comment by u12206050** [comment]
Any update on this layout? Otherwise I'll be starting to create my own one today! :)

**Reply by madc** [reply]
Would love to support if you decide to build it publicly.

**Comment by madc** [comment]
Just discovered [[directus-extensions-dcb-tree-view-bundle](https://github.com/pticon91/directus-extensions-dcb-tree-view-bundle)](https://github.com/pticon91/directus-extensions-dcb-tree-view-bundle/) by @pticon91. Wasn't able to get it running in a quick test session, but it look really cool and well thought through.

**Reply by digital-cowboy-91** [reply]
Hi Matt, could you provide me some details about that quick test session? Yesterday I've installed the extension from Directus Marketplace to my instance running in Digital Ocean and had no issue.

**Reply by madc** [reply]
First, thank you for your work on this extension, I'm really looking forward to use it!
Yesterday I just spend 15 Min trying to integrate it into an existing project and it did not work right away. I'll have another go and report back in the next couple of days.

**Reply by guidomallee** [reply]
This seems really promising. Unfortunately, I didn't manage to get it working in my Directus cloud instance. Probably because my collection id is a UUID, by the looks of the error message:
<img width="1239" alt="image" src="https://github.com/directus/directus/assets/15421432/c1ad4e59-06c3-4c6a-9929-6eb6224f042d">

**Reply by digital-cowboy-91** [reply]
Hi @guidomallee , could we move this conversation here please: https://github.com/pticon91/directus-extensions-dcb-tree-view-bundle/issues/1

**Comment by MentalGear** [comment]
# 🥇 [directus-extensions-dcb-tree-view-bundle](https://github.com/pticon91/directus-extensions-dcb-tree-view-bundle)

Just checked out the extension, and I think it might be exactly what you'd hope for in a tree layout - including level-based slug generation - makes organizing content so much cleaner, intuitive and allows to have just 1 Source-of-Truth.

Huge thanks to @digital-cowboy-91 for the fantastic work! I genuinely think this should be integrated into the core—it’s that useful!

**Reply by MentalGear** [reply]
One more honourable mention: https://github.com/codihaus/directus-extension-hierarchy-layout

**Comment by MentalGear** [comment]
## 🚀  Official: Tree View Table Layout (finally!)

It seems like there is now an official directus "Tree View Layout" extension - released just a few days ago.
Here's hoping it will soon be added to the view options as default!

You can find it in the Market Place > **"Tree View Table Layout"**.

**Reply by MentalGear** [reply]
Unfortunately I was unable to make the official "Tree View Table Layout" work - even though I followed the instructions - I think (I find the readme  somewhat lacking in clarity, maybe they could get inspired by the easiness of the above mentioned 3rd party solutions).

Since there's no repo, I'm leaving this as feedback and hopefully @ukmadlz @bryantgillespie @phazonoverload could provide more insights.


https://github.com/user-attachments/assets/dfa89593-4cce-47ba-b197-a7df30716149

**Reply by phazonoverload** [reply]
Beating us to our own announcement on Tuesday 😉 repo is [directus-labs/extensions](https://github.com/directus-labs/extensions).

**Reply by MentalGear** [reply]
Again, a much appreciated addition! Found the issue: it's related to a rather wide bug (you might want to get fixed before release).
The Extension only works if the table's primary ID is set to auto-increment. UUID or custom strings do not work.
Previous Bug Report: https://github.com/directus-labs/extensions/issues/115

**Reply by ComfortablyCoding** [reply]
> Again, a much appreciated addition! Found the issue: it's related to a rather wide bug (you might want to get fixed before release). The Extension only works if the table's primary ID is set to auto-increment. UUID or custom strings do not work. Previous Bug Report: [directus-labs/extensions#115](https://github.com/directus-labs/extensions/issues/115)

Thanks for the investigation/confirmation! The above issue should be resolved once https://github.com/directus-labs/extensions/pull/123 is merged/deployed

**Reply by MentalGear** [reply]
Thank you for your quick response on this, @ComfortablyCoding  & @formfcw !

I’ve been wondering how to leverage the Tree Layout structure that the CMS users can build on the application/routing side as well. My goal is for the Layout to act as a Single Source of Truth for organizing and structuring collection items.

For instance, if I have a collection of pages, I’d like to connect **routing paths** directly to the Tree Layout View’s structure. This might include handling of potential issues like duplicate item names.

Example page collection items / routes:
`/about`
`/projects/projectID/about`

In this scenario, both webpages would have corresponding items in the pages collection titled "about."

I’d appreciate your thoughts or suggestions on how best to approach:
a) connecting Tree Layout to routing (directus CMS querying approach)
b) handling duplicate item names/properties used for routing

PS: you might want to get inspired by the excellent display options & auto slug generation of https://github.com/digital-cowboy-91/directus-extensions-dcb-tree-view-bundle?tab=readme-ov-file#6-autogenerate-slug--path

**Reply by smkthat** [reply]
🤝 Hello @ComfortablyCoding & @formfcw !

How can I enable manual sorting for a user who does not have administrator rights? Manual sorting becomes available only if the "Admin Access" flag is set in the Access Policy.

A very useful extension! Thanks for the great work!

https://github.com/user-attachments/assets/b0d73ec5-896a-420c-9855-484a3f27db62

**Reply by formfcw** [reply]
Hey @smkthat! Thank you for your comment. There is an existing [PR](https://github.com/directus-labs/extensions/pull/216) for it. I will have a look as soon as possible.

**Comment by TorbenHammes** [comment]
> ## 🚀  Official: Tree View Table Layout (finally!)
> 
> 
> 
> It seems like there is now an official directus "Tree View Layout" extension - released just a few days ago.
> 
> Here's hoping it will soon be added to the view options as default!
> 
> 
> 
> You can find it in the Market Place > **"Tree View Table Layout"**.

That's great, but saving the layout is still missing.  When I change pages, I want to be able to either save the change or, if changed by mistake, cancel it. 

It would also be good to be able to edit slugs when moving elements. Slugs already exist as an extension, but children are not included there.

Perhaps the previous Tree View extension in this thread can be considered here.

**Reply by MentalGear** [reply]
Agreed, the 3rd party extension does better at the moment.
